src/vs/workbench/api/browser/viewsExtensionPoint.ts

```diff
@@ -226,19 +236,76 @@ const viewsExtensionPoint: IExtensionPoint<ViewExtensionPointType> = ExtensionsR
 	jsonSchema: viewsContribution
 });
 
+
+export interface IViewResolver {
```

Comment:
@sandy081 Here's an explanation of the changes I made in the views file:

1. View contributions now have an optional `type`. This can either be `tree` or `webview`.

    It would be great if we could determine this automatically for extension authors but I wasn't able to find a clean way to do so.

1. The `viewsExtensionPoint` can now generate tree views and webview views.

     It selects which view to use based on the `type`

1. In order to avoid a cycle, I added the new `IViewResolverRegistry`.

    This is a registry that maps view types to resolvers. A resolver takes a `IViewDescriptor` and fills in the missing details, such as the specific type of view to use.

1. I updated `viewExtensionPoint` to register the resolver for tree views.

    This resolver should do the exact same thing `viewExtensionPoint` previously had hardcoded

1. The webview then registers a new resolver for webview based views. 


You know this code much better than I do so let me know if you have other suggestions on how this could be accomplished. The main thing is that the `viewExtensionPoint` cannot import any webview code since webviews are at a higher level in the codebase

Reply:
> It would be great if we could determine this automatically for extension authors but I wasn't able to find a clean way to do so.

It is cool but I think it adds confusion and complexity. It is clear if extension declares a tree/web view and creates the same using API. 

**Note:** We need to validate if extension is creating the correct type of view for what it declared.

> In order to avoid a cycle, I added the new IViewResolverRegistry.
This is a registry that maps view types to resolvers. A resolver takes a IViewDescriptor and fills in the missing details, such as the specific type of view to use.

Not sure if I understood the need for this resolver and why there is a cycle? For web view type, you have to pass the `WebviewViewPane` to `ctorDescriptor` , may I know where we have cycle here?

CCing @alexr00 as she owns custom tree views

Reply:
I cleaned up the implementation so that this is no longer required. The changes in the view code should be much more minimal now